# KNN with scikit-learn - Lab

## Introduction

In this lab, you'll learn how to use scikit-learn's implementation of a KNN classifier on the classic Titanic dataset from Kaggle!
 

## Objectives

In this lab you will:

- Conduct a parameter search to find the optimal value for K 
- Use a KNN classifier to generate predictions on a real-world dataset 
- Evaluate the performance of a KNN model  


## Getting Started

Start by importing the dataset, stored in the `titanic.csv` file, and previewing it.

In [49]:
# Import pandas and set the standard alias
import pandas as pd

# Import the data from 'titanic.csv' and store it in a pandas DataFrame
raw_df = pd.read_csv('titanic.csv')

# Print the head of the DataFrame to ensure everything loaded correctly
print(raw_df.head())


   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


Great!  Next, you'll perform some preprocessing steps such as removing unnecessary columns and normalizing features.

## Preprocessing the data

Preprocessing is an essential component in any data science pipeline. It's not always the most glamorous task as might be an engaging data visual or impressive neural network, but cleaning and normalizing raw datasets is very essential to produce useful and insightful datasets that form the backbone of all data powered projects. This can include changing column types, as in: 


```python
df['col_name'] = df['col_name'].astype('int')
```
Or extracting subsets of information, such as: 

```python
import re
df['street'] = df['address'].map(lambda x: re.findall('(.*)?\n', x)[0])
```

> **Note:** While outside the scope of this particular lesson, **regular expressions** (mentioned above) are powerful tools for pattern matching! See the [regular expressions official documentation here](https://docs.python.org/3.6/library/re.html). 

Since you've done this before, you should be able to do this quite well yourself without much hand holding by now. In the cells below, complete the following steps:

1. Remove unnecessary columns (`'PassengerId'`, `'Name'`, `'Ticket'`, and `'Cabin'`) 
2. Convert `'Sex'` to a binary encoding, where female is `0` and male is `1` 
3. Detect and deal with any missing values in the dataset:  
    * For `'Age'`, replace missing values with the median age for the dataset  
    * For `'Embarked'`, drop the rows that contain missing values
4. One-hot encode categorical columns such as `'Embarked'` 
5. Store the target column, `'Survived'`, in a separate variable and remove it from the DataFrame  

While we always want to worry about data leakage, which is why we typically perform the split before the preprocessing, for this data set, we'll do some of the preprocessing first. The reason for this is that some of the values of the variables only have a handful of instances, and we want to make sure we don't lose any of them.

In [50]:
# Drop the unnecessary columns
raw_df = raw_df.drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin'])
raw_df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [51]:
# Fill missing values with the mode (most common value), which is 'male' in this case
raw_df['Sex'].fillna('male', inplace=True)
# Convert 'Sex' to binary encoding
raw_df['Sex'] = raw_df['Sex'].map({'female': 0, 'male': 1})
raw_df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    1  22.0      1      0   7.2500        S
1         1       1    0  38.0      1      0  71.2833        C
2         1       3    0  26.0      0      0   7.9250        S
3         1       1    0  35.0      1      0  53.1000        S
4         0       3    1  35.0      0      0   8.0500        S

In [52]:
# Find the number of missing values in each column
raw_df.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Embarked      2
dtype: int64

In [53]:
# Impute the missing values in 'Age'
raw_df['Age'].fillna(raw_df['Age'].median(), inplace=True)
raw_df.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    2
dtype: int64

In [54]:
# Drop the rows missing values in the 'Embarked' column
raw_df.dropna(subset=['Embarked'], inplace=True)

df = raw_df
df.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
SibSp       0
Parch       0
Fare        0
Embarked    0
dtype: int64

In [55]:
# One-hot encode the categorical columns
one_hot_df = pd.get_dummies(raw_df, columns=['Embarked'])
one_hot_df.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         0       3    1  22.0      1      0   7.2500           0           0   
1         1       1    0  38.0      1      0  71.2833           1           0   
2         1       3    0  26.0      0      0   7.9250           0           0   
3         1       1    0  35.0      1      0  53.1000           0           0   
4         0       3    1  35.0      0      0   8.0500           0           0   

   Embarked_S  
0           1  
1           0  
2           1  
3           1  
4           1

In [56]:
# Assign the 'Survived' column to labels
labels = raw_df['Survived']

# Drop the 'Survived' column from one_hot_df
one_hot_df.drop(columns=['Survived'])

Pclass  Sex   Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
0         3    1  22.0      1      0   7.2500           0           0   
1         1    0  38.0      1      0  71.2833           1           0   
2         3    0  26.0      0      0   7.9250           0           0   
3         1    0  35.0      1      0  53.1000           0           0   
4         3    1  35.0      0      0   8.0500           0           0   
..      ...  ...   ...    ...    ...      ...         ...         ...   
886       2    1  27.0      0      0  13.0000           0           0   
887       1    0  19.0      0      0  30.0000           0           0   
888       3    0  28.0      1      2  23.4500           0           0   
889       1    1  26.0      0      0  30.0000           1           0   
890       3    1  32.0      0      0   7.7500           0           1   

     Embarked_S  
0             1  
1             0  
2             1  
3             1  
4             1  
..          ...  
886           1  
887           1  
888           1  
889           0  
890           0  

[889 rows x 9 columns]

## Create training and test sets

Now that you've preprocessed the data, it's time to split it into training and test sets. 

In the cell below:

* Import `train_test_split` from the `sklearn.model_selection` module 
* Use `train_test_split()` to split the data into training and test sets, with a `test_size` of `0.25`. Set the `random_state` to 42 

In [57]:
from sklearn.model_selection import train_test_split

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(one_hot_df, labels, test_size=0.25, random_state=42)

# Print the shapes of the resulting datasets to verify
print("Training set shape:", X_train.shape)
print("Test set shape:", X_test.shape)
print("Training labels shape:", y_train.shape)
print("Test labels shape:", y_test.shape)

Training set shape: (666, 10)
Test set shape: (223, 10)
Training labels shape: (666,)
Test labels shape: (223,)


## Normalizing the data

The final step in your preprocessing efforts for this lab is to **_normalize_** the data. We normalize **after** splitting our data into training and test sets. This is to avoid information "leaking" from our test set into our training set (read more about data leakage [here](https://machinelearningmastery.com/data-leakage-machine-learning/) ). Remember that normalization (also sometimes called **_Standardization_** or **_Scaling_**) means making sure that all of your data is represented at the same scale. The most common way to do this is to convert all numerical values to z-scores. 

Since KNN is a distance-based classifier, if data is in different scales, then larger scaled features have a larger impact on the distance between points.

To scale your data, use `StandardScaler` found in the `sklearn.preprocessing` module. 

In the cell below:

* Import and instantiate `StandardScaler` 
* Use the scaler's `.fit_transform()` method to create a scaled version of the training dataset  
* Use the scaler's `.transform()` method to create a scaled version of the test dataset  
* The result returned by `.fit_transform()` and `.transform()` methods will be numpy arrays, not a pandas DataFrame. Create a new pandas DataFrame out of this object called `scaled_df`. To set the column names back to their original state, set the `columns` parameter to `one_hot_df.columns` 
* Print the head of `scaled_df` to ensure everything worked correctly 

In [61]:
# Import StandardScaler
from sklearn.preprocessing import StandardScaler

# Instantiate StandardScaler
scaler = StandardScaler()

# Transform the training and test sets
scaled_data_train = X_train_scaled = scaler.fit_transform(X_train)
scaled_data_test = X_test_scaled = scaler.transform(X_test)

# Convert into a DataFrame
scaled_df_train = pd.DataFrame(X_train_scaled, columns=X_train.columns)
scaled_df_test = pd.DataFrame(X_test_scaled, columns=X_test.columns)
scaled_df_train.head()
scaled_df_test.head()

Survived    Pclass       Sex       Age     SibSp     Parch      Fare  \
0 -0.795206  0.815528  0.719086 -0.120137 -0.474917 -0.480663 -0.488077   
1  1.257535 -1.587755 -1.390655 -1.183062  0.413551  1.972376  1.745013   
2  1.257535  0.815528 -1.390655 -1.183062  0.413551 -0.480663 -0.420624   
3 -0.795206 -0.386113  0.719086  0.031710 -0.474917 -0.480663 -0.385612   
4  1.257535 -1.587755 -1.390655 -0.879369 -0.474917  1.972376  0.941548   

   Embarked_C  Embarked_Q  Embarked_S  
0   -0.483046   -0.311768    0.620174  
1   -0.483046   -0.311768    0.620174  
2    2.070197   -0.311768   -1.612452  
3   -0.483046   -0.311768    0.620174  
4   -0.483046   -0.311768    0.620174

You may have noticed that the scaler also scaled our binary/one-hot encoded columns, too! Although it doesn't look as pretty, this has no negative effect on the model. Each 1 and 0 have been replaced with corresponding decimal values, but each binary column still only contains 2 values, meaning the overall information content of each column has not changed.

## Fit a KNN model

Now that you've preprocessed the data it's time to train a KNN classifier and validate its accuracy. 

In the cells below:

* Import `KNeighborsClassifier` from the `sklearn.neighbors` module 
* Instantiate the classifier. For now, you can just use the default parameters  
* Fit the classifier to the training data/labels
* Use the classifier to generate predictions on the test data. Store these predictions inside the variable `test_preds` 

In [63]:
# Import KNeighborsClassifier
from sklearn.neighbors import KNeighborsClassifier

# Instantiate KNeighborsClassifier
clf = KNeighborsClassifier()

# Fit the classifier
clf.fit(scaled_df_train, y_train)

# Predict on the test set
test_preds = clf.predict(scaled_df_test)

# Print the first few predictions to verify
print(test_preds[:10])

[0 1 1 0 1 1 0 0 1 1]


## Evaluate the model

Now, in the cells below, import all the necessary evaluation metrics from `sklearn.metrics` and complete the `print_metrics()` function so that it prints out **_Precision, Recall, Accuracy, and F1-Score_** when given a set of `labels` (the true values) and `preds` (the models predictions). 

Finally, use `print_metrics()` to print the evaluation metrics for the test predictions stored in `test_preds`, and the corresponding labels in `y_test`. 

In [64]:
# Your code here 
# Import the necessary functions
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score

In [65]:
def print_metrics(labels, preds):
    print("Precision Score: {}".format(precision_score(labels, preds)))
    print("Recall Score: {}".format(recall_score(labels, preds)))
    print("Accuracy Score: {}".format(accuracy_score(labels, preds)))
    print("F1 Score: {}".format(f1_score(labels, preds)))
    
print_metrics(y_test, test_preds)

Precision Score: 0.9634146341463414
Recall Score: 0.9634146341463414
Accuracy Score: 0.9730941704035875
F1 Score: 0.9634146341463414


> Interpret each of the metrics above, and explain what they tell you about your model's capabilities. If you had to pick one score to best describe the performance of the model, which would you choose? Explain your answer.

Write your answer below this line: 


________________________________________________________________________________

**Precision Score: 0.963**
*Interpretation*: Precision measures the proportion of positive predictions that are correct. A precision score of 0.963 means that 96.3% of the predictions made by the model for the positive class are accurate.
*Implication*: This high precision indicates that the model is very effective at avoiding false positives. In other words, when it predicts survival, it's almost always correct.

**Recall Score: 0.963**
*Interpretation*: Recall measures the proportion of actual positive cases that the model correctly identifies. A recall score of 0.963 means that the model identifies 96.3% of all actual survivors.
*Implication*: This high recall indicates that the model is very good at identifying most of the survivors in the dataset. It captures nearly all the true positives.

**Accuracy Score: 0.973**
*Interpretation*: Accuracy measures the overall proportion of correct predictions (both positives and negatives) out of all predictions. An accuracy of 0.973 means that the model correctly classifies 97.3% of all instances.
*Implication*: This high accuracy suggests that the model performs very well overall, making correct predictions in the majority of cases.

**F1 Score: 0.963**
*Interpretation*: The F1 Score is the harmonic mean of precision and recall. An F1 score of 0.963 indicates that the balance between precision and recall is very good, with both metrics being high.
*Implication*: This score reflects a balance between precision and recall, showing that the model performs well in both aspects without favoring one over the other.


## Improve model performance

While your overall model results should be better than random chance, they're probably mediocre at best given that you haven't tuned the model yet. For the remainder of this notebook, you'll focus on improving your model's performance. Remember that modeling is an **_iterative process_**, and developing a baseline out of the box model such as the one above is always a good start. 

First, try to find the optimal number of neighbors to use for the classifier. To do this, complete the `find_best_k()` function below to iterate over multiple values of K and find the value of K that returns the best overall performance. 

The function takes in six arguments:
* `X_train`
* `y_train`
* `X_test`
* `y_test`
* `min_k` (default is 1)
* `max_k` (default is 25)
    
> **Pseudocode Hint**:
1. Create two variables, `best_k` and `best_score`
1. Iterate through every **_odd number_** between `min_k` and `max_k + 1`. 
    1. For each iteration:
        1. Create a new `KNN` classifier, and set the `n_neighbors` parameter to the current value for k, as determined by the loop 
        1. Fit this classifier to the training data 
        1. Generate predictions for `X_test` using the fitted classifier 
        1. Calculate the **_F1-score_** for these predictions 
        1. Compare this F1-score to `best_score`. If better, update `best_score` and `best_k` 
1. Once all iterations are complete, print the best value for k and the F1-score it achieved 

In [66]:
def find_best_k(X_train, y_train, X_test, y_test, min_k=1, max_k=25):
    # Your code here
    best_k = None
    best_score = -1  # Start with a very low score

    for k in range(min_k, max_k + 1, 2):  # Iterate over odd values of k
        # Instantiate and fit the KNN classifier
        knn = KNeighborsClassifier(n_neighbors=k)
        knn.fit(X_train, y_train)
        
        # Generate predictions
        preds = knn.predict(X_test)
        
        # Calculate the F1-score
        score = f1_score(y_test, preds)
        
        # Update best_k and best_score if the current score is better
        if score > best_score:
            best_k = k
            best_score = score
    
    print("Best K: {}".format(best_k))
    print("Best F1 Score: {}".format(best_score))


In [67]:
find_best_k(scaled_data_train, y_train, scaled_data_test, y_test)
# Expected Output:

# Best Value for k: 17
# F1-Score: 0.7468354430379746

Best K: 1
Best F1 Score: 0.975609756097561


If all went well, you'll notice that model performance has improved by 3 percent by finding an optimal value for k. For further tuning, you can use scikit-learn's built-in `GridSearch()` to perform a similar exhaustive check of hyperparameter combinations and fine tune model performance. For a full list of model parameters, see the [sklearn documentation !](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html)



## (Optional) Level Up: Iterating on the data

As an optional (but recommended!) exercise, think about the decisions you made during the preprocessing steps that could have affected the overall model performance. For instance, you were asked to replace the missing age values with the column median. Could this have affected the overall performance? How might the model have fared if you had just dropped those rows, instead of using the column median? What if you reduced the data's dimensionality by ignoring some less important columns altogether?

In the cells below, revisit your preprocessing stage and see if you can improve the overall results of the classifier by doing things differently. Consider dropping certain columns, dealing with missing values differently, or using an alternative scaling function. Then see how these different preprocessing techniques affect the performance of the model. Remember that the `find_best_k()` function handles all of the fitting; use this to iterate quickly as you try different strategies for dealing with data preprocessing! 

## Summary

Well done! In this lab, you worked with the classic Titanic dataset and practiced fitting and tuning KNN classification models using scikit-learn! As always, this gave you another opportunity to continue practicing your data wrangling skills and model tuning skills using Pandas and scikit-learn!